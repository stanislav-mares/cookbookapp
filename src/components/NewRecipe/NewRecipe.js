import React from 'react';
import PropTypes from 'prop-types';

import AvatarCreator from 'components/shared/AvatarCreator/AvatarCreator';

export default class NewRecipe extends React.Component {

  constructor(props) {
    super(props);

    this.state = {
      title: this.props.title ? this.props.title : '',
      desc: this.props.desc ? this.props.desc : '',
      ingredients: this.props.ingredients ? this.props.ingredients.join('/') : '',
      procedure: this.props.procedure ? this.props.procedure.join('/') : '',
      avatarURL: this.props.avatarURL ? this.props.avatarURL : ''
    };

    this.inputMessage = '';
  }

  //handle input of simple data types
  handleChange = (event) => {
    this.inputMessage = '';
    const target = event.target;
    const name = target.name;
    let value = target.value;

    this.setState({
      [name]: value
    });
  }

  //check if input is in correct format -> return true, else false
  checkInput = (data) => {

    const pattern = /[^a-zA-Z0-9,\.\-\/:\(\)]+/;

    if (data.length === 0 || pattern.test(data))
      return false;

    return true;
  }

  handleSubmit = (event) => {
    event.preventDefault();

    let wrongInputs = [];
    for (const key of Object.keys(this.state)) {
      if(!this.checkInput(this.state[key])) {
          if (key !== 'avatarURL')
            wrongInputs.push(key);
      }
    }

    if (wrongInputs.length === 0) {
      const newRecipe = {
        ...this.state,
        procedure: this.state.procedure.trim().split(/[/]+/),
        ingredients: this.state.ingredients.trim().split(/[/]+/),
        likes: 0,
        dislikes: 0
      }

      this.props.handleRecipe(newRecipe);
      this.inputMessage = 'Success!';

      this.setState({
        title: '',
        desc: '',
        procedure: '',
        ingredients: ''
      });
    }else {

      const errorLabel = {
        title: 'title',
        desc: 'description',
        ingredients: 'ingredients',
        procedure: 'procedure'
      };

      wrongInputs = wrongInputs.map((input) => errorLabel[input]);

      this.inputMessage = `Wrong inputs: ${wrongInputs} (Empty string or forbidden character)`;
      this.forceUpdate();
    }
  }

  toggleForm = () => {
    this.props.toggleNewRecipeForm();
  }

  setAvatarURL = (avatarURL) => {
    this.setState({
      avatarURL: avatarURL
    });
  }

  render() {
    return(
      <div>
        <form method="get" onSubmit={this.handleSubmit}>
          <label>Title</label>
          <input type="text" name="title" placeholder='enter title'
            onChange={this.handleChange} value={this.state.title}
          />
          <label>Description</label>
          <input type="text" name="desc" placeholder='enter description'
            onChange={this.handleChange}
            value={this.state.desc}
          />
          <label>Ingredients</label>
          <textarea type="text" name="ingredients" placeholder='enter ingredients (separating with new line)'
            onChange={this.handleChange}
            value={this.state.ingredients}
          >
          </textarea>
          <label>Procedure</label>
          <textarea type="text" name="procedure" placeholder='enter procedure (separating with new line)'
            onChange={this.handleChange}
            value={this.state.procedure}
          >
          </textarea>
          <AvatarCreator setAvatarURL={this.setAvatarURL}/>
          <p>{this.inputMessage}</p>
          <div>
            <button type="submit">Submit</button>
            <button onClick={this.toggleForm}>Cancel</button>
          </div>
        </form>
      </div>
    );
  }
}

NewRecipe.propTypes = {
  handleRecipe: PropTypes.func,
  toggleNewRecipeForm: PropTypes.func
};
