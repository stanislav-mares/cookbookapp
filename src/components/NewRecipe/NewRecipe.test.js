import React from 'react';
import enzymeConfig from 'testConfig/enzymeConfig';
import { shallow } from 'enzyme';

import NewRecipe from 'components/NewRecipe/NewRecipe';

describe('<NewRecipe />', () => {
  it('Testing if handleChange method setting state correctly', () => {
    const component = shallow(<NewRecipe />);
    const instance =  component.instance();

    const inputs = component.findWhere(node => {
      node.type() === 'input' || node.type() === 'textarea';
    });

    inputs.forEach((node, index) => {
        node.simulate('change', {target:{name:node.props().name, value:'abcd'}});
        expect(instance.state[node.props().name]).toBe('abcd');
    });
  });

  it('Testing checkInput method if works correctly', () => {
      const component = shallow(<NewRecipe />);
      const instance = component.instance();

      const correctChars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ123456789.-:()';
      const wrongChars = ['@', '#', '$', '%', '^', '&', '*', '_', '+',
                         '{', '}', '|', '?', '>', '<', '[', ']','"'];

      expect(instance.checkInput(correctChars)).toBe(true);
      wrongChars.forEach((item, i) => {
          expect(instance.checkInput(item)).toBe(false);
      });
  });

  it('Testing structure of the NewRecipe component', () => {
    const component = shallow(<NewRecipe />);

    expect(component).toMatchSnapshot();
  });
});
