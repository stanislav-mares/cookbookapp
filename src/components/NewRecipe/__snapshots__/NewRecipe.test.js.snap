// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewRecipe /> Testing structure of the NewRecipe component 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NewRecipe />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": <form
        method="get"
        onSubmit={[Function]}
      >
        <label>
          Title
        </label>
        <input
          name="title"
          onChange={[Function]}
          placeholder="enter title"
          type="text"
          value=""
        />
        <label>
          Description
        </label>
        <input
          name="desc"
          onChange={[Function]}
          placeholder="enter description"
          type="text"
          value=""
        />
        <label>
          Ingredients
        </label>
        <textarea
          name="ingredients"
          onChange={[Function]}
          placeholder="enter ingredients (separating with new line)"
          type="text"
          value=""
        />
        <label>
          Procedure
        </label>
        <textarea
          name="procedure"
          onChange={[Function]}
          placeholder="enter procedure (separating with new line)"
          type="text"
          value=""
        />
        <AvatarCreator
          setAvatarURL={[Function]}
        />
        <p>
          
        </p>
        <div>
          <button
            type="submit"
          >
            Submit
          </button>
          <button
            onClick={[Function]}
          >
            Cancel
          </button>
        </div>
      </form>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label>
            Title
          </label>,
          <input
            name="title"
            onChange={[Function]}
            placeholder="enter title"
            type="text"
            value=""
          />,
          <label>
            Description
          </label>,
          <input
            name="desc"
            onChange={[Function]}
            placeholder="enter description"
            type="text"
            value=""
          />,
          <label>
            Ingredients
          </label>,
          <textarea
            name="ingredients"
            onChange={[Function]}
            placeholder="enter ingredients (separating with new line)"
            type="text"
            value=""
          />,
          <label>
            Procedure
          </label>,
          <textarea
            name="procedure"
            onChange={[Function]}
            placeholder="enter procedure (separating with new line)"
            type="text"
            value=""
          />,
          <AvatarCreator
            setAvatarURL={[Function]}
          />,
          <p>
            
          </p>,
          <div>
            <button
              type="submit"
            >
              Submit
            </button>
            <button
              onClick={[Function]}
            >
              Cancel
            </button>
          </div>,
        ],
        "method": "get",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Title",
          },
          "ref": null,
          "rendered": "Title",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "name": "title",
            "onChange": [Function],
            "placeholder": "enter title",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Description",
          },
          "ref": null,
          "rendered": "Description",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "name": "desc",
            "onChange": [Function],
            "placeholder": "enter description",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Ingredients",
          },
          "ref": null,
          "rendered": "Ingredients",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "name": "ingredients",
            "onChange": [Function],
            "placeholder": "enter ingredients (separating with new line)",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "textarea",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "Procedure",
          },
          "ref": null,
          "rendered": "Procedure",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "name": "procedure",
            "onChange": [Function],
            "placeholder": "enter procedure (separating with new line)",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "textarea",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {
            "setAvatarURL": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                type="submit"
              >
                Submit
              </button>,
              <button
                onClick={[Function]}
              >
                Cancel
              </button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Submit",
                "type": "submit",
              },
              "ref": null,
              "rendered": "Submit",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Cancel",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Cancel",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": <form
          method="get"
          onSubmit={[Function]}
        >
          <label>
            Title
          </label>
          <input
            name="title"
            onChange={[Function]}
            placeholder="enter title"
            type="text"
            value=""
          />
          <label>
            Description
          </label>
          <input
            name="desc"
            onChange={[Function]}
            placeholder="enter description"
            type="text"
            value=""
          />
          <label>
            Ingredients
          </label>
          <textarea
            name="ingredients"
            onChange={[Function]}
            placeholder="enter ingredients (separating with new line)"
            type="text"
            value=""
          />
          <label>
            Procedure
          </label>
          <textarea
            name="procedure"
            onChange={[Function]}
            placeholder="enter procedure (separating with new line)"
            type="text"
            value=""
          />
          <AvatarCreator
            setAvatarURL={[Function]}
          />
          <p>
            
          </p>
          <div>
            <button
              type="submit"
            >
              Submit
            </button>
            <button
              onClick={[Function]}
            >
              Cancel
            </button>
          </div>
        </form>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label>
              Title
            </label>,
            <input
              name="title"
              onChange={[Function]}
              placeholder="enter title"
              type="text"
              value=""
            />,
            <label>
              Description
            </label>,
            <input
              name="desc"
              onChange={[Function]}
              placeholder="enter description"
              type="text"
              value=""
            />,
            <label>
              Ingredients
            </label>,
            <textarea
              name="ingredients"
              onChange={[Function]}
              placeholder="enter ingredients (separating with new line)"
              type="text"
              value=""
            />,
            <label>
              Procedure
            </label>,
            <textarea
              name="procedure"
              onChange={[Function]}
              placeholder="enter procedure (separating with new line)"
              type="text"
              value=""
            />,
            <AvatarCreator
              setAvatarURL={[Function]}
            />,
            <p>
              
            </p>,
            <div>
              <button
                type="submit"
              >
                Submit
              </button>
              <button
                onClick={[Function]}
              >
                Cancel
              </button>
            </div>,
          ],
          "method": "get",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Title",
            },
            "ref": null,
            "rendered": "Title",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "name": "title",
              "onChange": [Function],
              "placeholder": "enter title",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Description",
            },
            "ref": null,
            "rendered": "Description",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "name": "desc",
              "onChange": [Function],
              "placeholder": "enter description",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Ingredients",
            },
            "ref": null,
            "rendered": "Ingredients",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "name": "ingredients",
              "onChange": [Function],
              "placeholder": "enter ingredients (separating with new line)",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "textarea",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Procedure",
            },
            "ref": null,
            "rendered": "Procedure",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "name": "procedure",
              "onChange": [Function],
              "placeholder": "enter procedure (separating with new line)",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "textarea",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "class",
            "props": Object {
              "setAvatarURL": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "",
            },
            "ref": null,
            "rendered": "",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <button
                  type="submit"
                >
                  Submit
                </button>,
                <button
                  onClick={[Function]}
                >
                  Cancel
                </button>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "children": "Submit",
                  "type": "submit",
                },
                "ref": null,
                "rendered": "Submit",
                "type": "button",
              },
              Object {
                "instance": null,
                "key": null,
                "nodeType": "host",
                "props": Object {
                  "children": "Cancel",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Cancel",
                "type": "button",
              },
            ],
            "type": "div",
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
